R/get_undata.R

```diff
@@ -1,114 +1,327 @@
 
-  table <- polite::scrape(bow, query) %>%
-    rvest::html_element(css = ".DataContainer") %>%
-    rvest::html_table()
+  # specify csv download format
+  query[["Format"]] <- "csv"
+
+  # replace query shortcuts
+  names(query)[names(query) =="f"] <- "DataFilter"
+  names(query)[names(query) =="d"] <- "DataMartId"
+
+  # create full url for 'DownloadHandler'
+  # Found by:
+  #   1. In Chrome open 'View' -> 'Developer' -> 'Developer Tools'.
+  #   2. Open Network tab.
+  #   3. Click download button.
+  #   4. Copy url in new row added to Network table.
+  downloader_url <- httr::modify_url(downloader_url, query=query)
+  downloader_url <- utils::URLdecode(downloader_url)
+
+  # get a temporary location to download the zip file to
+  zipdir <- tempdir()
+  zipfile <- tempfile(tmpdir = zipdir)
 
-  # TODO: create page number column
+  # download the zip file
+  res <- httr::GET(downloader_url, httr::write_disk(zipfile))
+  stopifnot(res$status_code == 200)
 
-  # TODO: merge on footnotes
+  # get the name of the csv file within the zip file
+  fname <- utils::unzip(zipfile, list = TRUE)$Name
+  stopifnot(length(fname) == 1)
 
-  checkmate::assert_tibble(table)
-  return(table)
+  # unzip the file and move it to 'output_path'
+  utils::unzip(zipfile, exdir = zipdir)
+  fs::file_move(fs::path(zipdir, fname), output_path)
+  fs::file_delete(zipfile)
+
+  return(invisible(output_path))
+}
+
+#' @title Combine together downloaded undata files
+#'
+#' @param fpaths \[`character()`\]\cr
+#'   Path to separate undata files to combine together. Output of
+#'   `download_undata_table`.
+#' @param n_footnote_col \[`numeric(1)`\]\cr
+#'   Column number of the 'data' table where the `footnote_col` is also stored.
+#'   Default is '2'.
+#' @param data_footnoteid_col \[`character(1)`\]\cr
+#'   Name of column in the data table storing the id number. Default is
+#'   'Value Footnotes'.
+#' @param footnoteid_col \[`character(1)`\]\cr
+#'   Name of column in the footnote table storing the id number. Default is
+#'   'footnoteSeqID'.
+#' @param footnote_col \[`character(1)`\]\cr
+#'   Name of column in the footnote table storing the text. Default is 'Footnote'.
+#'
+#' @return \[`data.table(1)`\] with all data from input paths combined together.
+#'
+#' @details undata tables contain the standard dataset included in the online tables
+#'   plus sometimes a table at the bottom containing footnotes. This function
+#'
+#' @family undata
+#'
+#' @examples
+#' \dontrun{
+#' data <- read_undata_files
+#' }
+#' @export
+read_undata_files <- function(fpaths,
+                              n_footnote_col = 2,
+                              data_footnoteid_col = "Value Footnotes",
+                              footnoteid_col = "footnoteSeqID",
+                              footnote_col = "Footnote") {
+
+  if (!requireNamespace("readr", quietly = TRUE)) {
+    stop("Package \"readr\" needed for this function to work. Please install it.",
+         call. = FALSE)
+  }
```

Comment:
Is this needed? I think all singularities we use have the package at this point?

Reply:
This is to start reducing the number of mandatory dependencies we have. Listed here https://github.com/ihmeuw-demographics/demUtils/blob/master/DESCRIPTION and difference explained here https://r-pkgs.org/description.html#dependencies. We were starting to get warnings in this package about too many dependencies which we want to reduce. When we install `demUtils` in our containers, it automatically installs both the imports and suggested packages